diff --git a/scipy/io/matlab/_mio5_params.py b/scipy/io/matlab/_mio5_params.py
index 0d60b8e7a..dd80ec921 100644
--- a/scipy/io/matlab/_mio5_params.py
+++ b/scipy/io/matlab/_mio5_params.py
@@ -278,4 +278,4 @@ class MatlabOpaque(np.ndarray):
 
 
 OPAQUE_DTYPE = np.dtype(
-    [('s0', 'O'), ('s1', 'O'), ('s2', 'O'), ('arr', 'O')])
+    [('_Metadata', 'O'), ('_TypeSystem', 'O'), ('_Class', 'O')])
diff --git a/scipy/io/matlab/_mio5_utils.pyx b/scipy/io/matlab/_mio5_utils.pyx
index b7abd7104..bfe92c3aa 100644
--- a/scipy/io/matlab/_mio5_utils.pyx
+++ b/scipy/io/matlab/_mio5_utils.pyx
@@ -596,8 +596,7 @@ cdef class VarReader5:
         # all miMATRIX types except the mxOPAQUE_CLASS have dims and a
         # name.
         if mc == mxOPAQUE_CLASS:
-            header.name = None
-            header.dims = None
+            header.name = self.read_int8_string()
             return header
         header.n_dims = self.read_into_int32s(header.dims_ptr, sizeof(header.dims_ptr))
         if header.n_dims > _MAT_MAXDIMS:
@@ -721,8 +720,6 @@ cdef class VarReader5:
             process = 0
         elif mc == mxOPAQUE_CLASS:
             arr = self.read_opaque(header)
-            arr = mio5p.MatlabOpaque(arr)
-            # to make them more re-writeable - don't squeeze
             process = 0
         # ensure we have read checksum.
         read_ok = self.cstream.all_data_read()
@@ -984,10 +981,10 @@ cdef class VarReader5:
         # Neither res nor the return value of this function are cdef'd as
         # cnp.ndarray, because that only adds useless checks with current
         # Cython (0.23.4).
-        res = np.empty((1,), dtype=OPAQUE_DTYPE)
+        res = np.empty((1, ), dtype=OPAQUE_DTYPE)
         res0 = res[0]
-        res0['s0'] = self.read_int8_string()
-        res0['s1'] = self.read_int8_string()
-        res0['s2'] = self.read_int8_string()
-        res0['arr'] = self.read_mi_matrix()
+        res0["_TypeSystem"] = PyUnicode_FromString(self.read_int8_string())
+        res0["_Class"] = PyUnicode_FromString(self.read_int8_string())
+        res0["_Metadata"] = self.read_mi_matrix()
+
         return res
